### PR TITLE
JIT: Run single EH region repair pass after layout

### LINKED ISSUE #108608: Assertion failed 'bbNumTryLast <= bbNumOuterTryLast' during 'Optimize layout'
Repro on windows/x64:

```c#
// Found by Antigen
// Reduced from 246.98 KB to 7.53 KB.


using System;
using System.Collections.Generic;
using System.Runtime.CompilerServices;
using System.Runtime.Intrinsics;
using System.Runtime.Intrinsics.Arm;
using System.Runtime.Intrinsics.X86;
using System.Numerics;
public class TestClass
{
    public struct S2
    {
    }
    static char s_char_5 = 'J';
    static int s_int_9 = 5;
    static Vector128<ushort> s_v128_ushort_21 = Vector128<ushort>.Zero;
    static Vector128<float> s_v128_float_26 = Vector128.CreateScalar(-4.955224f);
    static Vector256<ushort> s_v256_ushort_31 = Vector256<ushort>.Zero;
    string string_73 = "TIKBS";
    S2 s2_122 = new S2();
    static int s_loopInvariant = 9;
    private static List<string> toPrint = new List<string>();
    [MethodImpl(MethodImplOptions.NoInlining)]
    public char LeafMethod2()
    {
        unchecked
        {
            return s_char_5;
        }
    }
    public void Method0()
    {
        unchecked
        {
            bool bool_173 = false;
            int int_179 = 1;
            switch (15/4)
            {
                case -5:
                {
                    break;
                }
                default:
                {
                    switch (s_char_5 = LeafMethod2())
                    {
                        case 'V':
                        {
                            try
                            {
                                for (; int_179 < 15*4; int_179++)
                                {
                                    switch (s_int_9 -= s_int_9)
                                    {
                                        case -1:
                                        {
                                            break;
                                        }
                                        case -2:
                                        {
                                            break;
                                        }
                                        case 5:
                                        {
                                            break;
                                        }
                                        case 1:
                                        {
                                            break;
                                        }
                                        case -2147483647:
                                        {
                                            switch ("A9EET" + "F1"+ (string_73 = "A9EET" + "F1")+ (string_73 = "A9EET" + "F1"+ (string_73 = "A9EET" + "F1")))
                                            {
                                                case "T0X4TEN23":
                                                {
                                                    break;
                                                }
                                                case "2G8A35G88":
                                                {
                                                    break;
                                                }
                                                case "A9EET":
                                                {
                                                    try
                                                    {
                                                        Vector512.AsSingle(15*4* Vector512<long>.AllBitsSet);
                                                    }
                                                    finally
                                                    {
                                                        Vector256.WithUpper(s_v256_ushort_31 += s_v256_ushort_31, s_v128_ushort_21);
                                                    }
                                                    break;
                                                }
                                                case "F1":
                                                {
                                                    break;
                                                }
                                                case "S":
                                                {
                                                    break;
                                                }
                                                case "IXTF":
                                                {
                                                    break;
                                                }
                                                case "O9FXYE":
                                                {
                                                    break;
                                                }
                                                default:
                                                {
                                                    break;
                                                }
                                            }
                                            break;
                                        }
                                        case -5:
                                        {
                                            if (Sse.CompareScalarUnorderedGreaterThan(s_v128_float_26 -= s_v128_float_26- (s_v128_float_26 += s_v128_float_26), s_v128_float_26 -= s_v128_float_26- (s_v128_float_26 += s_v128_float_26)))
                                            {
                                            }
                                            else
                                            {
                                            }
                                            break;
                                        }
                                        case 2147483646:
                                        {
                                            break;
                                        }
                                        case 0:
                                        {
                                            break;
                                        }
                                        default:
                                        {
                                            try
                                            {
                                                s2_122 = s2_122;
                                            }
                                            catch (System.MissingMemberException)
                                            {
                                            }
                                            break;
                                        }
                                    }
                                }
                            }
                            finally
                            {
                                if (bool_173 = 15>4)
                                {
                                    int __loopvar13 = s_loopInvariant, __loopSecondaryVar13_0 = 15-4;
                                }
                                else
                                {
                                }
                            }
                            break;
                        }
                        default:
                        {
                            break;
                        }
                    }
                    break;
                }
            }
            return;
        }
    }
    public static void Main(string[] args)
    {
    }
    public static int Antigen()
    {
        new TestClass().Method0();
        return string.Join(Environment.NewLine, toPrint).GetHashCode();
    }
}
/*
Environment:

set DOTNET_ReadyToRun=1
set DOTNET_JitObjectStackAllocation=1
set DOTNET_JitRandomOnStackReplacement=10
set DOTNET_JitVTableProfiling=1
set DOTNET_TC_OnStackReplacement=1
set DOTNET_TC_PartialCompilation=1
set DOTNET_TC_QuickJitForLoops=1
set DOTNET_TieredCompilation=0
set DOTNET_TieredPGO=1
set DOTNET_JitStress=2
set DOTNET_JitThrowOnAssertionFailure=1
set DOTNET_LegacyExceptionHandling=1

Debug: 1631896427

Release: 0
JIT assert failed:
Assertion failed 'bbNumTryLast <= bbNumOuterTryLast' in 'TestClass:Method0():this' during 'Optimize layout' (IL size 635; hash 0x46e9aa75; FullOpts)

    File: D:\a\_work\1\s\src\coreclr\jit\jiteh.cpp Line: 3193


*/

```

### PR DESCRIPTION
**Pull request by amanasifkhalid**
Fixes #108608. To simplify EH maintenance logic, leave EH regions in a deconstructed phase during block reordering, and correct the end block pointers of each region in a pass over the block list after. By walking the list backwards, we can short-circuit finding the end of each region, and we don't run into the issue of an EH region not being able to determine if it's at the end of its parent region due to some awkward placement of a sibling region.

### CODE REVIEW

**Comment by amanasifkhalid** [comment]
[No diffs.](https://dev.azure.com/dnceng-public/public/_build/results?buildId=830409&view=ms.vss-build-web.run-extensions-tab) [Antigen](https://dev.azure.com/dnceng-public/public/_build/results?buildId=830585&view=ms.vss-build-web.run-extensions-tab) and [Fuzzlyn](https://dev.azure.com/dnceng-public/public/_build/results?buildId=830586&view=ms.vss-build-web.run-extensions-tab) runs didn't hit anything related. Outerloop failures are #104927.

cc @dotnet/jit-contrib, @AndyAyersMS PTAL. This doesn't look like a surgical fix, but I think it's preferable to trying to patch up the edge cases of the old implementation. That approach felt pattern-matchy and thus fragile, and keeping the temporary end block of a parent EH region up-to-date during moves in `fgMoveColdBlocks` incurred diffs, since we use these temporary end pointers as insertion points for cold blocks. The diffs looked like an improvement, but we probably want to avoid codegen churn for .NET 9 at this point, and I'm planning on making `fgMoveColdBlocks` more EH-agnostic for .NET 10 anyway. I hope the simplicity of the new approach justifies its case for .NET 9, though I'm open to suggestions for more focused fixes -- either way, this is the approach I'm thinking of taking for .NET 10.

File: src/coreclr/jit/jiteh.cpp [hunk 1]
```diff
@@ -1253,6 +1253,89 @@ void Compiler::fgSetHndEnd(EHblkDsc* handlerTab, BasicBlock* newHndLast)
...
 }
 
+//-------------------------------------------------------------
+// fgFindEHRegionEnds: Walk the block list, and set each try/handler region's end block.
+//
+void Compiler::fgFindEHRegionEnds()
+{
+    assert(compHndBBtabCount != 0);
+    bool* const tryEndsSet   = new (this, CMK_Generic) bool[compHndBBtabCount * 2]{};
+    bool* const hndEndsSet   = tryEndsSet + compHndBBtabCount;
```
[the remark this answers is not in the document]
Comment (AndyAyersMS) [reply src/coreclr/jit/jiteh.cpp:1263]: You could instead to walk the EH table first and null out `ebdTryLast` and `ebdHndLast`, then use the null values there as your "not yet set" indicators?

**Review by AndyAyersMS — COMMENTED** [review]
When you run diffs you also need to look at all the zero-sized diffs to see if there are any EH clause diffs.

I don't know if you can see those in the CI reports.

File: src/coreclr/jit/jiteh.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (amanasifkhalid) [reply src/coreclr/jit/jiteh.cpp:1263]: Good point. I'm guessing it's faster/easier to vectorize zeroing out an array versus nulling out each clause's end pointers, though the reduction in local state seems ideal.

**Comment by amanasifkhalid** [comment]
> When you run diffs you also need to look at all the zero-sized diffs to see if there are any EH clause diffs.

I ran asmdiffs locally, and I didn't see anything reported, so I'm assuming this is truly a no-diff change? If we saw a change in the EH clauses due to different block placement, we'd see codegen diffs too, and if we saw the clauses change without any blocks moving, then that would probably be a correctness issue and trigger asserts, right? In other words, for a given block layout, there's only one correct set of EH clauses, so any diffs for the latter are wrong.

**Comment by AndyAyersMS** [comment]
Maybe another way of putting it -- do you see SPMI diffs on the repro case (with release compilers)? Presumably you should? Or else does the release JIT get the right results anyway?

**Comment by amanasifkhalid** [comment]
> Maybe another way of putting it -- do you see SPMI diffs on the repro case (with release compilers)? Presumably you should? Or else does the release JIT get the right results anyway?

I see. With Checked compilers, I can see the EH clauses are now correct post-layout. Here's the problematic one:
```
---------------------------------------------------------------------------------------------------------------------------------------------------------------------
BBnum BBid ref try hnd preds           weight   [IL range]   [jump]                            [EH region]        [flags]
---------------------------------------------------------------------------------------------------------------------------------------------------------------------
BB01 [0000]  1                             1    [000..019)-> BB02(0.5),BB11(0.5)     ( cond )                     i
BB02 [0037]  1       BB01                  8    [0B3..0CF)-> BB03(0.5),BB11(0.5)     ( cond )                     i bwd
BB03 [0003]  1  2    BB02                  2    [01A..01E)-> BB04(1),BB06(0)         ( cond ) T2      try {       i keep loophead hascall bwd
BB04 [0042]  2  2    BB03,BB06             2    [01E..???)-> BB05(1)                 (always) T2                  i keep hascall bwd
BB05 [0014]  4  2    BB13,BB05,BB04,BB08 256    [026..07E)-> BB07(0.5),BB05(0.5)     ( cond ) T2                  i loophead nullcheck bwd bwd-src
BB06 [0044]  1  2    BB03                  0    [01E..01E)-> BB04(1)                 (always) T2                  rare internal
BB07 [0011]  1  1    BB05                128    [???..???)-> BB08(1),BB09(0)         ( cond ) T1      try {       i keep hascall bwd
BB08 [0045]  2  1    BB07,BB09           128    [057..???)-> BB05(1)                 (always) T1                  i keep hascall bwd
BB09 [0047]  1  1    BB07                  0    [???..???)-> BB08(1)                 (always) T1      } }         rare internal
BB10 [0008]  0  0                          0    [03E..04F)                           (throw ) T0      try { }     i rare keep bwd
BB11 [0041]  2       BB01,BB02             1    [019..0D0)                           (return)                     i
+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++ funclets follow
BB12 [0009]  1  2  0                       0    [04F..057)                           (throw ) T2 H0 F fault { }   i rare keep flet nullcheck bwd
BB13 [0012]  1  2  1                       0    [061..064)-> BB05(1)                 ( cret ) T2 H1 F catch { }   i rare keep xentry flet bwd
BB14 [0016]  1     2                       0    [080..095)                           (falret)    H2 F fault { }   i rare keep xentry flet bwd
---------------------------------------------------------------------------------------------------------------------------------------------------------------------

***************  Exception Handling table
index  eTry, eHnd
  0  ::   2        - Try at BB10..BB10 [03E..04F), Fault   at BB12..BB12 [04F..057)
  1  ::   2        - Try at BB07..BB09 [058..061), Handler at BB13..BB13 [061..064)
  2  ::            - Try at BB03..BB09 [01B..080), Fault   at BB14..BB14 [080..095)
```

And the correct one (notice the end block for T2 being pushed back by one):
```
---------------------------------------------------------------------------------------------------------------------------------------------------------------------
BBnum BBid ref try hnd preds           weight   [IL range]   [jump]                            [EH region]        [flags]
---------------------------------------------------------------------------------------------------------------------------------------------------------------------
BB01 [0000]  1                             1    [000..019)-> BB02(0.5),BB11(0.5)     ( cond )                     i
BB02 [0037]  1       BB01                  8    [0B3..0CF)-> BB03(0.5),BB11(0.5)     ( cond )                     i bwd
BB03 [0003]  1  2    BB02                  2    [01A..01E)-> BB04(1),BB06(0)         ( cond ) T2      try {       i keep loophead hascall bwd
BB04 [0042]  2  2    BB03,BB06             2    [01E..???)-> BB05(1)                 (always) T2                  i keep hascall bwd
BB05 [0014]  4  2    BB13,BB05,BB04,BB08 256    [026..07E)-> BB07(0.5),BB05(0.5)     ( cond ) T2                  i loophead nullcheck bwd bwd-src
BB06 [0044]  1  2    BB03                  0    [01E..01E)-> BB04(1)                 (always) T2                  rare internal
BB07 [0011]  1  1    BB05                128    [???..???)-> BB08(1),BB09(0)         ( cond ) T1      try {       i keep hascall bwd
BB08 [0045]  2  1    BB07,BB09           128    [057..???)-> BB05(1)                 (always) T1                  i keep hascall bwd
BB09 [0047]  1  1    BB07                  0    [???..???)-> BB08(1)                 (always) T1      }           rare internal
BB10 [0008]  0  0                          0    [03E..04F)                           (throw ) T0      try { } }   i rare keep bwd
BB11 [0041]  2       BB01,BB02             1    [019..0D0)                           (return)                     i
+++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++++ funclets follow
BB12 [0009]  1  2  0                       0    [04F..057)                           (throw ) T2 H0 F fault { }   i rare keep flet nullcheck bwd
BB13 [0012]  1  2  1                       0    [061..064)-> BB05(1)                 ( cret ) T2 H1 F catch { }   i rare keep xentry flet bwd
BB14 [0016]  1     2                       0    [080..095)                           (falret)    H2 F fault { }   i rare keep xentry flet bwd
---------------------------------------------------------------------------------------------------------------------------------------------------------------------

***************  Exception Handling table
index  eTry, eHnd
  0  ::   2        - Try at BB10..BB10 [03E..04F), Fault   at BB12..BB12 [04F..057)
  1  ::   2        - Try at BB07..BB09 [058..061), Handler at BB13..BB13 [061..064)
  2  ::            - Try at BB03..BB10 [01B..080), Fault   at BB14..BB14 [080..095)
```

I don't think we can dump EH tables in Release builds by default, though I can tweak this locally and take a look. Are you asking about this to evaluate the risk of this bug in product builds?

**Comment by AndyAyersMS** [comment]
> I don't think we can dump EH tables in Release builds by default, though I can tweak this locally and take a look. Are you asking about this to evaluate the risk of this bug in product builds?

In part, yes. It is good at this stage to be able to say what might happen if we don't fix this.

**Comment by AndyAyersMS** [comment]
Also note SPMI doesn't rely on dumping, it is comparing the binary data the JIT sends back to the VM (which includes the EH info), so release mode diffs "work" but may be harder to understand.

**Comment by amanasifkhalid** [comment]
Got it, I created a collection of the repro, and replayed it with Release builds. I get a zero-sized diff, so without the fix, it looks like we're propagating the incorrect EH clause info to the VM. For this particular method, that doesn't seem to affect runtime, but I can imagine this breaking stack walking for methods that throw.

**Comment by amanasifkhalid** [comment]
@AndyAyersMS since we're seeing diffs with Release builds, do you think a .NET 9 fix is justified, or would you like to see a repro where stack walking breaks from this? I can try to put something together.

**Review by AndyAyersMS — APPROVED** [review]

**Comment by AndyAyersMS** [comment]
I'd say we should go ahead and backport, given that we know of a case where we got things wrong.

**Pull request merged**